contrib/src/msgpack/mod.rs

```diff
@@ -0,0 +1,131 @@
+extern crate rmp_serde;
+
+use std::ops::{Deref, DerefMut};
+use std::io::{Cursor, Read};
+
+use rocket::outcome::Outcome;
+use rocket::request::Request;
+use rocket::data::{self, Data, FromData};
+use rocket::response::{self, Responder, Response};
+use rocket::http::Status;
+
+use serde::{Serialize, Deserialize};
+
+pub use self::rmp_serde::decode::Error as MsgPackError;
+
+/// The `MsgPack` type: implements `FromData` and `Responder`, allowing you to easily
+/// consume and respond with MessagePack data.
+///
+/// If you're receiving MessagePack data, simply add a `data` parameter to your route
+/// arguments and ensure the type of the parameter is a `MsgPack<T>`, where `T` is
+/// some type you'd like to parse from MessagePack. `T` must implement `Deserialize`
+/// from [Serde](https://github.com/serde-rs/serde). The data is parsed from the
+/// HTTP request body.
+///
+/// ```rust,ignore
+/// #[post("/users/", format = "application/msgpack", data = "<user>")]
+/// fn new_user(user: MsgPack<User>) {
+///     ...
+/// }
+/// ```
+/// You don't _need_ to use `format = "application/msgpack"`, but it _may_ be what
+/// you want. Using `format = application/msgpack` means that any request that
+/// doesn't specify "application/msgpack" as its first `Content-Type:` header
+/// parameter will not be routed to this handler. By default, Rocket will accept a
+/// Content Type of any of the following for MessagePack data:
+/// `application/msgpack`, `application/x-msgpack`, or `bin/msgpack`.
+///
+/// If you're responding with MessagePack data, return a `MsgPack<T>` type, where `T`
+/// implements `Serialize` from [Serde](https://github.com/serde-rs/serde). The
+/// content type of the response is set to `application/msgpack` automatically.
+///
+/// ```rust,ignore
+/// #[get("/users/<id>")]
+/// fn user(id: usize) -> MsgPack<User> {
+///     let user_from_id = User::from(id);
+///     ...
+///     MsgPack(user_from_id)
+/// }
+/// ```
+///
+#[derive(Debug)]
+pub struct MsgPack<T>(pub T);
+
+impl<T> MsgPack<T> {
+    /// Consumes the `MsgPack` wrapper and returns the wrapped item.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use rocket_contrib::MsgPack;
+    /// let string = "Hello".to_string();
+    /// let my_msgpack = MsgPack(string);
+    /// assert_eq!(my_msgpack.into_inner(), "Hello".to_string());
+    /// ```
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+/// Maximum size of MessagePack data is 1MB.
+/// TODO: Determine this size from some configuration parameter.
+const MAX_SIZE: u64 = 1048576;
+
+impl<T: Deserialize> FromData for MsgPack<T> {
+    type Error = MsgPackError;
+
+    fn from_data(request: &Request, data: Data) -> data::Outcome<Self, Self::Error> {
+        // Accepted content types are:
+        // `application/msgpack`, `application/x-msgpack`, and `bin/msgpack`
+        if !request.content_type().map_or(false, |ct|
```

Comment:
Looks like `match` needs the type to be exactly the same, so even though `UncasedAscii` implements `PartialEq` for `str`, Rust's typechecker won't allow this.

I did modify it slightly in a way that improves readability, although I don't think it's as nice as this.